beps/0001-notifications-system/README.md

```diff
@@ -147,6 +148,8 @@ The link is a relative or absolute URL. As an example, it can be used:
 - by an external system to request an action within an asynchronous task
 - by a BE plugin to provide link to other part of the Backstage UI (i.e. to the Catalog)
 
+The metadata is a flexible JSON like field, where an additional payload can be stored.
```

Comment:
```suggestion
The metadata is an opaque JSON field, where an additional payload can be stored. The format of this data is owned by the notification sender and is tied to the notification topic, i.e. notifications sent from the source source on the same topic should use a compatible metadata format. The primary purpose of this field is to allow for custom processing and rendering based on the additional metadata.
```

Suggesting to be a bit more strict about the format of the metadata field. Most likely it's enough to document the format/type, but I could also see it being useful to even have an adjacent metadata version field, or even allow for multiple metadata buckets of different types, e.g.

```ts
  metadata: Array<{
    type: string;
    value: JsonValue;
  }>
```

Over all I'd like us to be very careful about making sure that this data is easy to consume in an explicit way, while making it possible to evolve the data over time. If we go forward with the multiple metadata idea I'd event suggest that we move some of the existing fields into metadata, like icon, links, etc.

Reply:
Additionally, we need some way for the other plugins to consume these notifications. Sure they can hook into the notifications signal but I am not sure if that's the proper way to distribute who wants which notification in the frontend.

Reply:
@Rugvip How about to use `metadata` as `JsonObject` for multiple multiple metadata buckets?
The problem with the array in this model that querying it with knex is not really supported.

It still can be the same idea, where we can say something like:

```
metadata: {
  link: {
     type: 'string',
     value: 'https://github.com/backstage/backstage'
  },
  age: {
     type: 'number',
     value: 30
  }
}
```

And this i also don't like as a user who will work with this. 

The more convenient way might be to have 2 fields: one `metadata` and another `metadataSchema`.
Then it will look like

"metadata"
```
{
  link: 'https://github.com/backstage/backstage',
  age: : 30,
}
```

"metadataSchema"
```
{
  link: 'string',
  age: 'number',
}
```

Completely agree that "link", "icon", "etc." can be moved to "metadata".

Reply:
@drodil Having a field "type" in notifications (could be an extra field in metadata as well) can help with filtering notifications per custom notification module.

Reply:
@acierto PostgreSQL has excellent support for arrays - if you normalize the schema 😁, i.e. storing the metadata in a separate table.

With `type` I had something more strict in mind, things like `core.link`, `catalog.entity-ref`, `core.icon.v1`, etc.

Reply:
Ah ok, that what you have in mind, I was thinking to use `jsonb` as a type of field and for that I couldn't find a solution to query arrays.

Reply:
If to some up the idea:

We would like to have an extra table, i.e.
`notification_payalod_metadata` with columns:

type - string
value - jsonb
version - string

On code level side:

```
  metadata: Array<{
    type: string;
    value: JsonValue;
    version: string;
  }>
```

@Rugvip with having `type` more strict, do you mean that the types should have more context? 
As we can't/shouldn't validate the type content.

I see the last example as `core.icon.v1`, does that mean that you prefer to keep versioning in the type name instead of a separate field?

Reply:
@acierto yes to all those things 😁